Description: Fix URI patch on subsequent requests
 Use the full URI (including "usergroup" or path) as specified in --server for
 all requests during authentication instead of only the first one
Author: Stefan Bühler <source@stbuehler.de>
Origin: upstream, https://gitlab.com/openconnect/openconnect/-/merge_requests/560
Bug: https://gitlab.com/openconnect/openconnect/-/issues/737
Reviewed-by: Lee Garrett <debian@rocketjump.eu>
Last-Update: 2025-10-28
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
diff --git a/auth.c b/auth.c
index 317fc21..e245b48 100644
--- a/auth.c
+++ b/auth.c
@@ -455,13 +455,15 @@ static int parse_auth_node(struct openconnect_info *vpninfo, xmlNode *xml_node,
 
 			/* defaults for new XML POST */
 			form->method = strdup("POST");
-			form->action = strdup("/");
 
 			xmlnode_get_prop(xml_node, "method", &form->method);
 			xmlnode_get_prop(xml_node, "action", &form->action);
 
-			if (!form->method || !form->action ||
-			    strcasecmp(form->method, "POST") || !form->action[0]) {
+			/* - expect unset action (reuse current URL) or non-empty action="..."
+			 * - expect unset method (defaults to "POST") or explicit method="POST"
+			 */
+			if ((form->action && !form->action[0]) ||
+			    !form->method || strcasecmp(form->method, "POST")) {
 				vpn_progress(vpninfo, PRG_ERR,
 					     _("Cannot handle form method='%s', action='%s'\n"),
 					     form->method, form->action);
diff --git a/www/changelog.xml b/www/changelog.xml
index 49a50b3..1ba207f 100644
--- a/www/changelog.xml
+++ b/www/changelog.xml
@@ -15,7 +15,7 @@
 <ul>
    <li><b>OpenConnect HEAD</b>
      <ul>
-       <li><i>No changelog entries yet</i></li>
+       <li>Use the full URI (including "usergroup" or path) as specified in <tt>--server</tt> for all requests during authentication instead of only the first one (<a href="https://gitlab.com/openconnect/openconnect/-/merge_requests/560">!560</a>).</li>
      </ul><br/>
   </li>
   <li><b><a href="https://www.infradead.org/openconnect/download/openconnect-9.12.tar.gz">OpenConnect v9.12</a></b>
